pkg/operators/wasm/wasm.go

```diff
@@ -0,0 +1,235 @@
+	"time"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/tetratelabs/wazero"
+	wapi "github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/oci"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+const (
+	wasmObjectMediaType = "application/vnd.gadget.wasm.program.v1+binary"
+)
+
+type wasmOperator struct{}
+
+func (w *wasmOperator) Name() string {
+	return "wasm"
+}
+
+func (w *wasmOperator) Description() string {
+	return "handles wasm programs"
+}
+
+func (w *wasmOperator) InstantiateImageOperator(
+	gadgetCtx operators.GadgetContext,
+	desc ocispec.Descriptor,
+	paramValues api.ParamValues,
+) (
+	operators.ImageOperatorInstance, error,
+) {
+	return &wasmOperatorInstance{
+		desc:      desc,
+		gadgetCtx: gadgetCtx,
+		handleMap: map[uint32]any{},
+		logger:    gadgetCtx.Logger(),
+	}, nil
+}
+
+type wasmOperatorInstance struct {
+	desc      ocispec.Descriptor
+	rt        wazero.Runtime
+	gadgetCtx operators.GadgetContext
+	mod       wapi.Module
+
+	logger logger.Logger
+
+	// malloc function exported by the guest
+	guestMalloc wapi.Function
+
+	// Golang objects are exposed to the wasm module by using a handleID
+	handleMap  map[uint32]any
+	handleCtr  uint32
+	handleLock sync.RWMutex
+}
+
+func (i *wasmOperatorInstance) Name() string {
+	return "wasm"
+}
+
+func (i *wasmOperatorInstance) Prepare(gadgetCtx operators.GadgetContext) error {
+	err := i.init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm: %w", err)
+	}
+	err = i.Init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm guest: %w", err)
+	}
+
+	return nil
+}
+
+func (i *wasmOperatorInstance) ExtraParams(gadgetCtx operators.GadgetContext) api.Params {
+	return nil
+}
+
+func (i *wasmOperatorInstance) addHandle(obj any) uint32 {
+	if obj == nil {
+		return 0
+	}
+
+	i.handleLock.Lock()
+	defer i.handleLock.Unlock()
+	i.handleCtr++
+	if i.handleCtr == 0 { // 0 is reserved
+		i.handleCtr++
+	}
+	xctr := 0
+	for {
+		if xctr > 1<<32 {
+			// exhausted; TODO: report somehow
+			return 0
+		}
+		if _, ok := i.handleMap[i.handleCtr]; !ok {
+			// register new entry
+			i.handleMap[i.handleCtr] = obj
+			return i.handleCtr
+		}
+		xctr++
+	}
+}
+
+func (i *wasmOperatorInstance) getHandle(handleID uint32) any {
+	i.handleLock.RLock()
+	defer i.handleLock.RUnlock()
+	return i.handleMap[handleID]
+}
+
+func (i *wasmOperatorInstance) delHandle(handleID uint32) {
+	i.handleLock.Lock()
+	defer i.handleLock.Unlock()
+	delete(i.handleMap, handleID)
+}
+
+func (i *wasmOperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+	ctx := gadgetCtx.Context()
+	rtConfig := wazero.NewRuntimeConfig().
+		WithCloseOnContextDone(true).
+		WithMemoryLimitPages(256) // 16MB (64KB per page)
```

Comment:
Should be exposed as a global parameter (even if we don't initialize that properly for now).

Reply:
mmm, I see ImageOperators don't have a `GlobalParams` method. What are you suggesting for this one?

Reply:
I suggest making it, at least, a constant for now.

Reply:
Opened #3072